### PR TITLE
error handling, fix #2

### LINKED ISSUE #2: Test for accessible pcbnew API, prompt to update kicad if needed
Defensive check is better than a stack trace in user's face.

### PR DESCRIPTION
**Pull request by Andrei-Pozolotin**

### CODE REVIEW

File: plugin.py [hunk 1]
```diff
@@ -49,26 +49,50 @@ def prepare(self):
...
+            return False
+        
+    # invoke chain of dependent methods
+    def RunSteps(self):
+        return self.RunExport() and self.RunRouter() and self.RunImport()
 
+    # kicad plugin action entry
     def Run(self):
+        if not has_pcbnew_api():
+            show_error("Missing required python API:\n pcbnew.ExportSpecctraDSN\n pcbnew.ImportSpecctraSES\n")
```
Comment (qu1ck) [inline plugin.py]: I would add a note that plugin requires development nightly build from 2019-05-05 or later.

File: plugin.py [hunk 2]
```diff
@@ -77,5 +101,25 @@ def Run(self):
...
+
+
+# verify required pcbnew api is present
+def has_pcbnew_api():
+    return hasattr(pcbnew, 'ExportSpecctraDSN') and hasattr(pcbnew, 'ImportSpecctraSES')
+
+
+# report error to the user gui
+def show_error(message):
+    caption = "kicad_freerouting-plugin"
```
Comment (qu1ck) [inline plugin.py]: nit: "Freerouting plugin"
as user string it doesn't need to avoid spaces

File: plugin.py [hunk 3]
```diff
@@ -49,26 +49,50 @@ def prepare(self):
...
+    # kicad plugin action entry
     def Run(self):
+        if not has_pcbnew_api():
+            show_error("Missing required python API:\n pcbnew.ExportSpecctraDSN\n pcbnew.ImportSpecctraSES\n")
+            return
         self.prepare()
-        self.invoke(self.RunExport)
-        self.invoke(self.RunRouter)
-        self.invoke(self.RunImport)
+        safe_invoke(self.RunSteps)
```
Comment (qu1ck) [inline plugin.py]: It's not good to wait on java in wx gui thread. Limit what you run in safe_invoke to only what is necessary, in this case import.

**Review by qu1ck — COMMENTED** [review]
Thanks for the invite.
Few comments below.

File: plugin.py (hunk 1 above)
Comment (Andrei-Pozolotin) [reply plugin.py]: applied

File: plugin.py (hunk 2 above)
Comment (Andrei-Pozolotin) [reply plugin.py]: applied

File: plugin.py (hunk 3 above)
Comment (Andrei-Pozolotin) [reply plugin.py]: the intent is:
1. block any `pcbnew` state changes while `java` is active
2. invoke `self.RunSteps` as single unit of dependent chained methods

since `safe_invoke` is asynchronous, how would you do that?

**Comment by Andrei-Pozolotin** [comment]
> Thanks for the invite.

I have another idea I wanted to run by you.

I want to propose to "the community" to create common plugin repository
https://github.com/kicad-extra
and move all plugin projects there, so that all devs can own all projects, similar to
https://github.com/KiCad

what do you think?

File: plugin.py (hunk 3 above)
Comment (qu1ck) [reply plugin.py]: You can't do it properly with current API we have. 
Problem with blocking UI thread is that OS window manager events will not be processed either so you will get "Not responding" status with potentially more side effects.
At this point I think it's ok to leave it up to user: if you change the pcb while autorouting happens and then import breaks your changes it's on you.
Not running java in UI thread will also allow you to have a dialog with a cancel button if user changes his mind and doesn't want to wait for autorouter or import the result.

**Comment by qu1ck** [comment]
I don't think having common plugin repository will work well. People have too different coding standards, styles, processes. And it's ok but it won't work well in a single repo with shared ownership.

I had ideas about plugin manager in KiCad that would provide central discovery point as well as manage updates, compatibility checks. List of plugins showed there would be centrally managed by KiCad team and to apply to be listed a plugin would have to pass some conformity checks like have predefined format that plugin manager can parse to extract description, version info and other metadata.

I think that will work much better in the long run. Once I finalize my proposal I will present it on kicad dev list.

File: plugin.py (hunk 3 above)
Comment (Andrei-Pozolotin) [reply plugin.py]: please review a hack around this problem

**Comment by Andrei-Pozolotin** [comment]
>I don't think having common plugin repository will work well.

got it, thank you. one vote down :-) https://github.com/kicad-extra/arkon/issues/1

File: plugin.py [hunk 4]
```diff
@@ -46,77 +45,83 @@ def prepare(self):
...
             ---
             Try development nightly build:
             * http://kicad-pcb.org/download/
             """)
-            return
-        self.prepare()
-        safe_invoke(self.RunSteps)
 
 
 # provision gui-thread-safe execution context
```
Comment (qu1ck) [inline plugin.py:137]: You don't need this anymore and it actually prevents KiCad process from exiting.

File: plugin.py [hunk 5]
```diff
@@ -46,77 +45,83 @@ def prepare(self):
...
-                return False
-        except Exception as error:
-                show_error("""
-                Java process failure: 
-                %s""" % str(error))
+            else:
+                invoker.show_error()
                 return False
+        else:
+            return False  # should not happen
```
Comment (qu1ck) [inline plugin.py]: It's pretty idiomatic to write
```
if result == wx.ID_OK:
  ...
else:
  ...
```
Also it's best to destroy dialog here where you create it. No need to do it in CallAfter().

File: plugin.py [hunk 6]
```diff
@@ -46,77 +45,83 @@ def prepare(self):
...
-            output, error = process.communicate()
-            if process.returncode == 0:
+        command = self.module_command
+        dialog = wx_make_cancel("""
+        Complete or Terminate FreeRouting:
+        * to complete, close Java window
+        * to terminate, press Terminate here
+        """)
+        invoker = ProcessThread(command, dialog)
+        dialog.Show()  # dialog first
```
Comment (qu1ck) [inline plugin.py:81]: Why is this needed?

File: plugin.py (hunk 4 above)
Comment (Andrei-Pozolotin) [reply plugin.py:137]: I do not follow.
1. wx.CallAfter required for import board.ses
2. wx.App instance is required by wx.CallAfter

File: plugin.py (hunk 5 above)
Comment (Andrei-Pozolotin) [reply plugin.py]: > No need to do it in CallAfter()

tried that, it crashes kicad

File: plugin.py (hunk 6 above)
Comment (Andrei-Pozolotin) [reply plugin.py:81]: to avoid race condition, when `ProcessThread` is trying to destroy dialog which is not yet shown

**Comment by Andrei-Pozolotin** [comment]
I squashed the commits, perhaps it is easier to live test this pr now.

File: plugin.py (hunk 4 above)
Comment (qu1ck) [reply plugin.py:137]: You don't even reference theApp anywhere. And when you run in KiCad you already have a wxApp (it's c++ counterpart) initialized because kicad itself is a wxwidgets app. So wx.CallAfter works just fine.

File: plugin.py (hunk 5 above)
Comment (qu1ck) [reply plugin.py]: On windows this does not work. Dialog never closes even after java process completes.

What you would do normally do is call dialog.EndModal(resultcode) where you want to close it and destroy it right where you call dialog.ShowModal() to not leak it. 
Problem is wxMessageDialog is not actually a wxDialog so it ignores EndModal. 
To do this properly you will have to make an instance of wx.Dialog yourself.

File: plugin.py (hunk 4 above)
Comment (qu1ck) [reply plugin.py:137]: Also on linux because of extra wxApp created here the kicad process never exits cleanly and when you restart it you get "kicad is already running, are you sure?".

**Comment by qu1ck** [comment]
I did some testing on both linux and win. Both platforms have their issues with current implementation that I outlined in comments above.

I believe this should work:
1. Get rid of wx.App(). I tested this on both platforms, it doesn't break wx.CallAfter() when run from kicad.
2. Make terminate dialog actual wx.Dialog instead of wxMessageDialog. It's a bit of tedious work but can be simplified a lot with wxFormBuilder.
3. Close dialog in ProcessThread with EndModal, destroy right after ShowModal()
4. I think you need to join() on the thread after it's done. You may be getting away with it because it's daemon thread but it may leak resources if you don't.

**Comment by Andrei-Pozolotin** [comment]
great stuff, thank you. applying your ideas now.

**Comment by Andrei-Pozolotin** [comment]
please test

results so far:

> 1. Get rid of wx.App(). I tested this on both platforms, it doesn't break wx.CallAfter() when run from kicad.

wx.CallAfter still crashes here w/o wx.App. likely due to use of `phoenix`

build options:
https://github.com/random-builder/kicad_freerouting-plugin/blob/master/doc/build-options.md

current work around is based on:
 https://git.launchpad.net/kicad/tree/pcbnew/python/kicad_pyshell/__init__.py#n89

> 2. Make terminate dialog actual wx.Dialog instead of wxMessageDialog. It's a bit of tedious work but can be simplified a lot with wxFormBuilder.

works, except could not figure out how to auto-resize to fit text

> 3. Close dialog in ProcessThread with EndModal, destroy right after ShowModal()

works

wow, based on the diagram https://docs.wxwidgets.org/3.0/classwx_message_dialog.html
who would guess: `wxMessageDialog is not actually a wxDialog so it ignores EndModal` :-)

> 4. I think you need to join() on the thread after it's done. You may be getting away with it because it's daemon thread but it may leak resources if you don't.

works

File: plugin.py (hunk 4 above)
Comment (Andrei-Pozolotin) [reply plugin.py:137]: yes, see #1

**Comment by qu1ck** [comment]
I dug a bit deeper on the wx.App issue and found this:
https://github.com/KiCad/kicad-source-mirror/commit/8b060799ebcde69dab29565ca5dd63d36c6ac87b
which led me to this:
https://bugs.launchpad.net/kicad/+bug/1809913
and this:
https://github.com/wxWidgets/Phoenix/issues/1126

It boils down to this. You can't do gui stuff from non-gui thread so you need wx.CallAfter, you can't do CallAfter without wx.App so you create one. But then you bork the global wxApp instance reference in c++ land and Kicad event loop never finishes.

I guess not much can be done about it on plugin side at the moment so we will have to live with this.

I tried to move wx.App() init into plugin method so that new App instance would be garbage collected. Plugin works but that makes things way worse with random crashes in c++ code after plugin finishes :rofl: which is not unexpected given the above.

> wow, based on the diagram...

Yeah, misleading, I know :) I think it's because wxGtk (and probably wxMsw too) use native error boxes that just don't have the needed APIs to handle EndModal and some other things so it's ignored.

> works, except could not figure out how to auto-resize to fit text

Use wx.StaticText instead of TextCtrl and it will autosize properly on sizer.Fit().

**Comment by qu1ck** [comment]
I just had an idea that instead of trying to coerce gui thread to do what we need with CallAfter() we could do it the old fashioned way - by sending it an event. Some quick googling on the topic popped this great collection of the solutions:
https://wiki.wxpython.org/LongRunningTasks

Very first one there is detailing the solution I am talking about and it has a great code example too.

Keep the dialog around so that you have something to send an event to and treat the event handler as CallAfter() function. 
I'm pretty sure this would solve all the issues.

**Comment by Andrei-Pozolotin** [comment]
please review 

> I dug a bit deeper on the wx.App issue

cool, and @sethhillbrand seems to agree on the following work around 
https://bugs.launchpad.net/kicad/+bug/1827742
```
kicad should automatically provision global wx.App instance by default,
similar to how it is done in kicad_pyshell
```
> Use wx.StaticText instead of TextCtrl 

works, thank you

> I just had an idea that instead ... by sending it an event

ok, thinking. look at what `wx.CallAfter` is doing:
https://github.com/wxWidgets/wxPython/blob/master/src/_core_ex.py#L146

**Comment by Andrei-Pozolotin** [comment]
also: `wxAnyThread`
https://github.com/BrendanSimon/wxanythread/blob/master/wxAnyThread/__init__.py

**Comment by qu1ck** [comment]
> ok, thinking. look at what wx.CallAfter is doing

Yeah, it's doing same thing underneath but it's sending the event to App. You can avoid that if you send it to a dialog, that way you don't have to create your own App at all.
wxAnyThread seems to be using same idea. But if you choose to use that I suggest you copy that code into your repo instead of adding it as dependency. Installing such deps is bothersome on all platforms except linux.

Also feel free to commit this as is and update to event posting code later if you prefer. Change looks good to me.

**Comment by Andrei-Pozolotin** [comment]
> Yeah, it's doing same thing underneath ...

Result so far: 
* `wx.PostEvent` works for `ProcessThread` -> `ProcessDialog` notification
* any attempt to bypass `wx.App` instance crashes `kicad`

> Also feel free to commit this as is

Yes, lets do that for now.

> Change looks good

Thank you so much for your support!
